src/java/org/apache/nutch/util/SitemapProcessor.java

```diff
@@ -0,0 +1,436 @@
+ *     crawls and inject the urls from sitemap to the crawldb.</li>
+ * </ol>
+ *
+ * <p>For more details see:
+ *      https://wiki.apache.org/nutch/SitemapFeature </p>
+ */
+public class SitemapProcessor extends Configured implements Tool {
+  public static final Logger LOG = LoggerFactory.getLogger(SitemapProcessor.class);
+  public static final SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+
+  public static final String CURRENT_NAME = "current";
+  public static final String LOCK_NAME = ".locked";
+  public static final String SITEMAP_STRICT_PARSING = "sitemap.strict.parsing";
+  public static final String SITEMAP_URL_FILTERING = "sitemap.url.filter";
+  public static final String SITEMAP_URL_NORMALIZING = "sitemap.url.normalize";
+
+  private static class SitemapMapper extends Mapper<Text, Writable, Text, CrawlDatum> {
+    private ProtocolFactory protocolFactory = null;
+    private boolean strict = true;
+    private boolean filter = true;
+    private boolean normalize = true;
+    private URLFilters filters = null;
+    private URLNormalizers normalizers = null;
+    private CrawlDatum datum = new CrawlDatum();
+    private SiteMapParser parser = null;
+
+    public void setup(Context context) {
+      Configuration conf = context.getConfiguration();
+      this.protocolFactory = new ProtocolFactory(conf);
+      this.filter = conf.getBoolean(SITEMAP_URL_FILTERING, true);
+      this.normalize = conf.getBoolean(SITEMAP_URL_NORMALIZING, true);
+      this.strict = conf.getBoolean(SITEMAP_STRICT_PARSING, true);
+      this.parser = new SiteMapParser(strict);
+
+      if (filter)
+        filters = new URLFilters(conf);
+      if (normalize)
+        normalizers = new URLNormalizers(conf, URLNormalizers.SCOPE_DEFAULT);
+    }
+
+    public void map(Text key, Writable value, Context context) throws IOException, InterruptedException {
+      String url;
+
+      try {
+        if (value instanceof CrawlDatum) {
+          // If its an entry from CrawlDb, emit it. It will be merged in the reducer
+          context.write(key, (CrawlDatum) value);
+        }
+        else if (value instanceof HostDatum) {
+          // For entry from hostdb, get sitemap url(s) from robots.txt, fetch the sitemap,
+          // extract urls and emit those
+
+          // try different combinations of schemes one by one till we get rejection in all cases
+          String host = key.toString();
+          if((url = filterNormalize("http://" + host + "/")) == null &&
+              (url = filterNormalize("https://" + host + "/")) == null &&
+              (url = filterNormalize("ftp://" + host + "/")) == null &&
+              (url = filterNormalize("file:/" + host + "/")) == null) {
+            context.getCounter("Sitemap", "filtered_records").increment(1);
+            return;
+          }
+
+          BaseRobotRules rules = protocolFactory.getProtocol(url).getRobotRules(new Text(url), datum, new LinkedList<>());
```

Comment:
Always passing a new LinkedList as the third parameter to the [.getRobotsRules](https://builds.apache.org/job/nutch-trunk/javadoc/org/apache/nutch/protocol/Protocol.html#getRobotRules-org.apache.hadoop.io.Text-org.apache.nutch.crawl.CrawlDatum-java.util.List-) method call may not be preferable. I've looked at the code and we have the option to pass null. This needs to be tested.
I have seen elsewhere in the codebase that use of this signature aligns with use of fetcher.store.robotstxt configuration property... so we may wish to so the same here and align it.